File: skill-evaluator/scripts/security_scan.py
#!/usr/bin/env python3
"""
Security scanner for Agent Skills.
Scans scripts/ directory for high-risk patterns.

Usage:
    python security_scan.py /path/to/skill
    
Output:
    JSON with scan results
"""

import re
import os
import sys
import json
from pathlib import Path

HIGH_RISK_PATTERNS = {
    'python': [
        (r'\beval\s*\(', 'eval() - arbitrary code execution'),
        (r'\bexec\s*\(', 'exec() - arbitrary code execution'),
        (r'__import__\s*\(', 'dynamic import - potential code injection'),
        (r'compile\s*\([^)]+[\'"]exec[\'"]\s*\)', 'compile with exec mode'),
        (r'os\.system\s*\(', 'os.system() - shell command execution'),
        (r'subprocess\.call\s*\([^)]*shell\s*=\s*True', 'subprocess with shell=True'),
        (r'subprocess\.Popen\s*\([^)]*shell\s*=\s*True', 'Popen with shell=True'),
        (r'subprocess\.run\s*\([^)]*shell\s*=\s*True', 'subprocess.run with shell=True'),
        (r'[\'\"]/etc/', 'access to /etc/ system directory'),
        (r'[\'\"]/root/', 'access to /root/ directory'),
        (r'[\'\"]~/.ssh/', 'access to SSH keys'),
        (r'[\'\"]~/.aws/', 'access to AWS credentials'),
        (r'[\'\"]~/.config/', 'access to user config'),
        (r'\.env[\'\"\\s]', 'access to .env file'),
        (r'requests\.(post|put|patch)\s*\(\s*[\'\"](https?://(?!localhost|127\.0\.0\.1))', 'external data transmission'),
        (r'urllib\.request\.urlopen', 'URL open - potential SSRF'),
        (r'socket\.connect\s*\(', 'raw socket connection'),
        (r'base64\.b64decode\s*\([^)]+\).*exec', 'base64 decode with execution'),
        (r'pickle\.load', 'pickle.load - arbitrary code execution'),
        (r'yaml\.load\s*\([^)]*(?!Loader\s*=\s*yaml\.SafeLoader)', 'unsafe yaml.load'),
    ],
    'bash': [
        (r'rm\s+-rf\s+/', 'rm -rf on root path'),
        (r'rm\s+-rf\s+\*', 'rm -rf with wildcard'),
        (r'curl\s+[^\|]+\|\s*bash', 'curl piped to bash'),
        (r'wget\s+[^\|]+\|\s*bash', 'wget piped to bash'),
        (r'\beval\s+', 'bash eval command'),
        (r'cat\s+/etc/passwd', 'reading /etc/passwd'),
        (r'cat\s+/etc/shadow', 'reading /etc/shadow'),
        (r'cat\s+~/.ssh/id_', 'reading SSH private key'),
        (r'\$\([^)]*curl', 'command substitution with curl'),
        (r'`[^`]*curl', 'backtick execution with curl'),
    ],
    'javascript': [
        (r'\beval\s*\(', 'eval() - arbitrary code execution'),
        (r'new\s+Function\s*\(', 'Function constructor - code execution'),
        (r'child_process\.exec\s*\(', 'child_process.exec - shell command'),
        (r'child_process\.execSync\s*\(', 'child_process.execSync - shell command'),
        (r'require\s*\(\s*[\'\"]\s*child_process', 'importing child_process'),
        (r'fs\.readFileSync\s*\([^)]*(/etc/|~/.ssh|~/.aws)', 'reading sensitive files'),
    ]
}

MEDIUM_RISK_PATTERNS = {
    'python': [
        (r'ctypes\.', 'ctypes usage - native code'),
        (r'importlib\.', 'dynamic import'),
        (r'getattr\s*\([^,]+,\s*[^\'\"]+\)', 'getattr with dynamic attribute'),
    ],
    'bash': [
        (r'chmod\s+777', 'chmod 777 - insecure permissions'),
        (r'sudo\s+', 'sudo usage'),
    ],
    'javascript': [
        (r'require\s*\(\s*[a-zA-Z_][a-zA-Z0-9_]*\s*\)', 'dynamic require'),
    ]
}


def get_language(filepath):
    """Determine language from file extension."""
    ext = Path(filepath).suffix.lower()
    lang_map = {
        '.py': 'python',
        '.sh': 'bash',
        '.bash': 'bash',
        '.js': 'javascript',
        '.ts': 'javascript',
        '.mjs': 'javascript',
    }
    return lang_map.get(ext)


def scan_content(filepath, content, lang):
    """Scan file content for security patterns."""
    findings = []
    lines = content.split('\n')
    
    # Check high-risk patterns
    for pattern, desc in HIGH_RISK_PATTERNS.get(lang, []):
        for i, line in enumerate(lines, 1):
            if re.search(pattern, line, re.IGNORECASE):
                findings.append({
                    'file': filepath,
                    'line': i,
                    'pattern': desc,
                    'severity': 'HIGH',
                    'content': line.strip()[:100]
                })
    
    # Check medium-risk patterns
    for pattern, desc in MEDIUM_RISK_PATTERNS.get(lang, []):
        for i, line in enumerate(lines, 1):
            if re.search(pattern, line, re.IGNORECASE):
                findings.append({
                    'file': filepath,
                    'line': i,
                    'pattern': desc,
                    'severity': 'MEDIUM',
                    'content': line.strip()[:100]
                })
    
    return findings


def scan_skill(skill_path):
    """Scan a skill directory for security issues."""
    skill_dir = Path(skill_path)
    scripts_dir = skill_dir / 'scripts'
    
    results = {
        'skill_path': str(skill_dir),
        'scan_type': 'static',
        'files_scanned': 0,
        'findings': [],
        'summary': {
            'high': 0,
            'medium': 0,
            'low': 0
        }
    }
    
    if not scripts_dir.exists():
        results['note'] = 'No scripts/ directory found'
        results['passed'] = True
        return results
    
    for filepath in scripts_dir.rglob('*'):
        if not filepath.is_file():
            continue
            
        lang = get_language(str(filepath))
        if not lang:
            continue
        
        results['files_scanned'] += 1
        
        try:
            content = filepath.read_text(encoding='utf-8', errors='replace')
            findings = scan_content(
                str(filepath.relative_to(skill_dir)),
                content,
                lang
            )
            results['findings'].extend(findings)
        except Exception as e:
            results['findings'].append({
                'file': str(filepath.relative_to(skill_dir)),
                'error': str(e),
                'severity': 'LOW'
            })
    
    # Calculate summary
    for f in results['findings']:
        sev = f.get('severity', 'LOW').lower()
        if sev in results['summary']:
            results['summary'][sev] += 1
    
    # Determine pass/fail
    results['passed'] = results['summary']['high'] == 0
    
    return results


def main():
    if len(sys.argv) < 2:
        print("Usage: python security_scan.py /path/to/skill", file=sys.stderr)
        sys.exit(1)
    
    skill_path = sys.argv[1]
    
    if not os.path.exists(skill_path):
        print(f"Error: Path does not exist: {skill_path}", file=sys.stderr)
        sys.exit(1)
    
    results = scan_skill(skill_path)
    print(json.dumps(results, indent=2))
    
    sys.exit(0 if results['passed'] else 1)


if __name__ == '__main__':
    main()

File: skill-evaluator/scripts/security_scan.js
#!/usr/bin/env node
/**
 * Security scanner for Agent Skills.
 * Scans scripts/ directory for high-risk patterns.
 * 
 * Usage:
 *   node security_scan.js /path/to/skill
 *   
 * Output:
 *   JSON with scan results (to stdout)
 *   
 * Exit codes:
 *   0 = passed (no HIGH findings)
 *   1 = failed (HIGH findings found)
 *   2 = error
 *   
 * Spec: See references/security-scan-spec.md
 */

const fs = require('fs');
const path = require('path');

// Pattern definitions per language
const HIGH_RISK_PATTERNS = {
  python: [
    [/\beval\s*\(/i, 'eval() - arbitrary code execution'],
    [/\bexec\s*\(/i, 'exec() - arbitrary code execution'],
    [/__import__\s*\(/i, 'dynamic import - potential code injection'],
    [/compile\s*\([^)]+['"]exec['"]\s*\)/i, 'compile with exec mode'],
    [/os\.system\s*\(/i, 'os.system() - shell command execution'],
    [/subprocess\.call\s*\([^)]*shell\s*=\s*True/i, 'subprocess with shell=True'],
    [/subprocess\.Popen\s*\([^)]*shell\s*=\s*True/i, 'Popen with shell=True'],
    [/subprocess\.run\s*\([^)]*shell\s*=\s*True/i, 'subprocess.run with shell=True'],
    [/['"]\/etc\//i, 'access to /etc/ system directory'],
    [/['"]\/root\//i, 'access to /root/ directory'],
    [/['"]~\/\.ssh\//i, 'access to SSH keys'],
    [/['"]~\/\.aws\//i, 'access to AWS credentials'],
    [/['"]~\/\.config\//i, 'access to user config'],
    [/\.env['"\\s]/i, 'access to .env file'],
    [/requests\.(post|put|patch)\s*\(\s*['"](https?:\/\/(?!localhost|127\.0\.0\.1))/i, 'external data transmission'],
    [/urllib\.request\.urlopen/i, 'URL open - potential SSRF'],
    [/socket\.connect\s*\(/i, 'raw socket connection'],
    [/base64\.b64decode\s*\([^)]+\).*exec/i, 'base64 decode with execution'],
    [/pickle\.load/i, 'pickle.load - arbitrary code execution'],
    [/yaml\.load\s*\([^)]*(?!Loader\s*=\s*yaml\.SafeLoader)/i, 'unsafe yaml.load'],
  ],
  javascript: [
    [/\beval\s*\(/i, 'eval() - arbitrary code execution'],
    [/new\s+Function\s*\(/i, 'Function constructor - code execution'],
    [/child_process\.(exec|execSync)\s*\(/i, 'shell command execution'],
    [/child_process\.spawn\s*\([^)]*shell:\s*true/i, 'spawn with shell option'],
    [/require\s*\(\s*['"]child_process/i, 'importing child_process'],
    [/fs\.(readFileSync|readFile)\s*\([^)]*(?:\/etc\/|~\/\.ssh|~\/\.aws)/i, 'reading sensitive files'],
  ],
  bash: [
    [/rm\s+-rf\s+\//i, 'rm -rf on root path'],
    [/rm\s+-rf\s+\*/i, 'rm -rf with wildcard'],
    [/curl\s+[^|]+\|\s*(ba)?sh/i, 'curl piped to shell'],
    [/wget\s+[^|]+\|\s*(ba)?sh/i, 'wget piped to shell'],
    [/\beval\s+/i, 'bash eval command'],
    [/cat\s+\/etc\/passwd/i, 'reading /etc/passwd'],
    [/cat\s+\/etc\/shadow/i, 'reading /etc/shadow'],
    [/cat\s+~\/\.ssh\/id_/i, 'reading SSH private key'],
    [/\$\([^)]*curl/i, 'command substitution with curl'],
    [/`[^`]*curl/i, 'backtick execution with curl'],
  ]
};

const MEDIUM_RISK_PATTERNS = {
  python: [
    [/ctypes\./i, 'ctypes usage - native code'],
    [/importlib\./i, 'dynamic import'],
    [/yaml\.load\s*\([^)]*(?!Loader\s*=\s*yaml\.SafeLoader)/i, 'unsafe yaml.load'],
  ],
  javascript: [
    [/require\s*\(\s*[a-zA-Z_][a-zA-Z0-9_]*\s*\)/i, 'dynamic require'],
  ],
  bash: [
    [/chmod\s+777/i, 'chmod 777 - insecure permissions'],
    [/\bsudo\s+/i, 'sudo usage'],
  ]
};

/**
 * Detect language from file extension
 */
function getLanguage(filepath) {
  const ext = path.extname(filepath).toLowerCase();
  const langMap = {
    '.py': 'python',
    '.sh': 'bash',
    '.bash': 'bash',
    '.js': 'javascript',
    '.mjs': 'javascript',
    '.ts': 'javascript',
  };
  return langMap[ext] || null;
}

/**
 * Scan file content for security patterns
 */
function scanContent(filepath, content, lang) {
  const findings = [];
  const lines = content.split('\n');
  
  // Check high-risk patterns
  const highPatterns = HIGH_RISK_PATTERNS[lang] || [];
  for (const [pattern, desc] of highPatterns) {
    lines.forEach((line, idx) => {
      if (pattern.test(line)) {
        findings.push({
          file: filepath,
          line: idx + 1,
          pattern: desc,
          severity: 'HIGH',
          content: line.trim().slice(0, 100)
        });
      }
    });
  }
  
  // Check medium-risk patterns
  const mediumPatterns = MEDIUM_RISK_PATTERNS[lang] || [];
  for (const [pattern, desc] of mediumPatterns) {
    lines.forEach((line, idx) => {
      if (pattern.test(line)) {
        findings.push({
          file: filepath,
          line: idx + 1,
          pattern: desc,
          severity: 'MEDIUM',
          content: line.trim().slice(0, 100)
        });
      }
    });
  }
  
  return findings;
}

/**
 * Recursively get all files in directory
 */
function getAllFiles(dirPath, arrayOfFiles = []) {
  if (!fs.existsSync(dirPath)) {
    return arrayOfFiles;
  }
  
  const files = fs.readdirSync(dirPath);
  
  for (const file of files) {
    const fullPath = path.join(dirPath, file);
    if (fs.statSync(fullPath).isDirectory()) {
      getAllFiles(fullPath, arrayOfFiles);
    } else {
      arrayOfFiles.push(fullPath);
    }
  }
  
  return arrayOfFiles;
}

/**
 * Main scan function
 */
function scanSkill(skillPath) {
  const results = {
    skill_path: skillPath,
    scan_type: 'static',
    files_scanned: 0,
    findings: [],
    summary: {
      high: 0,
      medium: 0,
      low: 0
    }
  };
  
  const scriptsDir = path.join(skillPath, 'scripts');
  
  if (!fs.existsSync(scriptsDir)) {
    results.note = 'No scripts/ directory found';
    results.passed = true;
    return results;
  }
  
  const files = getAllFiles(scriptsDir);
  
  for (const filepath of files) {
    const lang = getLanguage(filepath);
    if (!lang) continue;
    
    results.files_scanned++;
    
    try {
      const content = fs.readFileSync(filepath, 'utf8');
      const relativePath = path.relative(skillPath, filepath);
      const findings = scanContent(relativePath, content, lang);
      results.findings.push(...findings);
    } catch (err) {
      results.findings.push({
        file: path.relative(skillPath, filepath),
        error: err.message,
        severity: 'LOW'
      });
    }
  }
  
  // Calculate summary
  for (const f of results.findings) {
    const sev = (f.severity || 'LOW').toLowerCase();
    if (sev in results.summary) {
      results.summary[sev]++;
    }
  }
  
  // Determine pass/fail
  results.passed = results.summary.high === 0;
  
  return results;
}

/**
 * Main entry point
 */
function main() {
  const args = process.argv.slice(2);
  
  if (args.length < 1) {
    console.error('Usage: node security_scan.js /path/to/skill');
    process.exit(2);
  }
  
  const skillPath = args[0];
  
  if (!fs.existsSync(skillPath)) {
    console.error(`Error: Path does not exist: ${skillPath}`);
    process.exit(2);
  }
  
  try {
    const results = scanSkill(skillPath);
    console.log(JSON.stringify(results, null, 2));
    process.exit(results.passed ? 0 : 1);
  } catch (err) {
    console.error(`Error: ${err.message}`);
    process.exit(2);
  }
}

main();
